libtenzir/builtins/operators/cache.cpp

```diff
@@ -84,131 +95,183 @@ struct cache_state {
-      self->quit(diagnostic::error("cache expired").to_error());
+    TENZIR_ASSERT(write_timeout_ > duration::zero());
+    on_write_timeout_.dispose();
+    on_write_timeout_ = self_->run_delayed_weak(write_timeout_, [this] {
+      self_->quit(diagnostic::error("cache expired").to_error());
     });
   }
 
   auto write_ok() -> caf::result<bool> {
-    return writer;
+    return writer_;
   }
 
   auto write(table_slice events) -> caf::result<bool> {
     TENZIR_ASSERT(events.rows() > 0);
-    const auto sender = self->current_sender();
+    const auto sender = self_->current_sender();
     TENZIR_ASSERT(sender);
-    if (not writer) {
-      writer = sender->address();
-      self->monitor(sender, [this](const caf::error&) {
-        TENZIR_ASSERT(not done);
-        done = true;
+    if (not writer_) {
+      writer_ = sender->address();
+      self_->monitor(sender, [this](const caf::error& err) {
+        TENZIR_ASSERT(not done_);
+        done_ = true;
+        update_multicaster_.push(cache_update{
+          .approx_bytes = byte_size_,
+          .state = err ? cache_state::failed : cache_state::closed,
+        });
+        update_multicaster_.close();
         reset_read_timeout();
         // We ignore error messages because they do not matter to the readers.
-        for (auto& [_, reader] : readers) {
-          if (reader.offset == cache.size() and reader.rp.pending()) {
+        for (auto& [_, reader] : readers_) {
+          if (reader.offset == cached_events_.size() and reader.rp.pending()) {
             reader.rp.deliver(table_slice{});
           }
         }
       });
       set_write_timeout();
-    } else if (writer != sender->address()) {
+    } else if (writer_ != sender->address()) {
       return false;
     }
     auto exceeded_capacity = false;
-    if (cache_size + events.rows() > capacity.inner) {
-      events = head(std::move(events), capacity.inner - cache_size);
+    if (cache_size_ + events.rows() > max_events_.inner) {
+      events = head(std::move(events), max_events_.inner - cache_size_);
       diagnostic::warning("cache exceeded capacity")
-        .primary(capacity.source)
-        .emit(diagnostics);
+        .primary(max_events_.source)
+        .emit(diagnostics_);
       exceeded_capacity = true;
       if (events.rows() == 0) {
         return false;
       }
     }
-    cache_size += events.rows();
-    cache.push_back(std::move(events));
-    for (auto& [_, reader] : readers) {
+    cache_size_ += events.rows();
+    // Calculate the byte size of any referenced Arrow buffers
+    // TODO: Use `table_slice::approx_bytes` instead.
+    byte_size_ += detail::narrow_cast<uint64_t>(
+      check(arrow::util::ReferencedBufferSize(*to_record_batch(events))));
+    // If a single cache exceeds the total capacity, we stop short of adding the
+    // batch of events that'd make it go over the limit. This is better than
+    // being evicted immediately.
```

Comment:
In that case we would not get all events when hitting the cache in a re-run of the same pipeline, right? Shouldn't we just discard the cache instead?

Reply:
Aldwin and I discussed that if a single cache is this large that it is generally better to keep around that one cache but truncate it (as if it had a capacity on its own) as opposed to throwing it away entirely by letting it run over the limit. The user gets a warning in this case in the pipeline writing into the cache.

Reply:
Is that really so obvious? That one cache is going to kick out all other caches that were created before, and once the next cache gets created this one will be kicked out too.

I believe when a pipeline emits a lot of data it often also means it emits that data at a high rate, and that means creating a single event is cheap. I think it would be better to cache events that are expensive to produce instead.

Reply:
This edge case only affects the situation where there exists a single cache. It slightly improves the handling in an already terrible situation by letting the user have one cache instead of zero.

Reply:
I have some ideas of how we can make the cache usage more effective, but we can do that in a follow up.